exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -19,44 +19,286 @@
+    // that there is no single implementation that can do it all well. So, we use multiple implementations
+    // chosen based on the pattern length.
+    if (patternLength == 0) {
+      matcherFcn = new MatcherZero();
+    } else if (patternLength == 1) {
+      matcherFcn = new MatcherOne();
+    } else if (patternLength == 2) {
+      matcherFcn = new MatcherTwo();
+    } else if (patternLength == 3) {
+      matcherFcn = new MatcherThree();
+    } else if (patternLength < 10) {
+      matcherFcn = new MatcherN();
+    } else {
+      matcherFcn = new BoyerMooreMatcher();
+    }
   }
 
   @Override
   public int match(int start, int end, DrillBuf drillBuf) {
+    return matcherFcn.match(start, end, drillBuf);
+  }
+
+  //--------------------------------------------------------------------------
+  // Inner Data Structure
+  // --------------------------------------------------------------------------
+
+  /** Abstract matcher class to allow us pick the most efficient implementation */
+  private abstract class MatcherFcn {
+    protected final byte[] patternArray;
+
+    protected MatcherFcn() {
+      assert patternByteBuffer.hasArray();
+
+      patternArray = patternByteBuffer.array();
+    }
+
+    /**
+     * @return 1 if the pattern was matched; 0 otherwise
+     */
+    protected abstract int match(int start, int end, DrillBuf drillBuf);
+  }
+
+  /** Handles patterns with length one */
+  private final class MatcherZero extends MatcherFcn {
 
-    if (patternLength == 0) { // Everything should match for null pattern string
+    private MatcherZero() {
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
       return 1;
     }
+  }
+
+  /** Handles patterns with length one */
+  private final class MatcherOne extends MatcherFcn {
+
+    private MatcherOne() {
+      super();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess = end - start;
+      final byte firstPattByte  = patternArray[0];
+
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != inputByte) {
+          continue;
+        }
+        return 1;
+      }
+      return 0;
+    }
+  }
+
+  /** Handles patterns with length two */
+  private final class MatcherTwo extends MatcherFcn {
+
+    private MatcherTwo() {
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess = end - start - 1;
+      final byte firstPattByte  = patternArray[0];
+      final byte secondPattByte = patternArray[1];
```

Comment:
can you initialize them in the constructor for matcher function ? that way you can initialize once and reuse for each match.

Reply:
Good idea; done this for all the matchers.